File: leetcode_alg/solution55.cpp
//
// Created by wangheng on 2021/10/10.
// 跳跃游戏

#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int k = 0;
        for(int i = 0;i<nums.size();i++){
            if(i>k)return false;
            k = max(k, i+nums[i]);
        }
        return true;
    }
};

int main() {
    auto s = new Solution;
    vector<int> nums = {2,3,1,1,4};
    cout<<s->canJump(nums)<<endl;
    return 0;
}

File: golang/solution/solution55.go
package solution

import "fmt"

// 跳跃游戏 - 从nums[0]点起跳，判断能否达到终点
func canJump(nums []int) bool {
	// 最远能跳的距离
	var k int = 0
	for i := 0; i < len(nums); i++ {
		if i > k {
			return false
		}
		k = max(k, nums[i]+i)
	}
	return true
}

func Solution55() {
	nums := []int{2, 3, 1, 1, 4}
	fmt.Println(canJump(nums))
}
